### DISCUSSION TITLE
Support for CUDA parallel programming

### DISCUSSION THREAD

**Discussion by gsaponaro** [robotology/community#10] [Q&A] [answered]
Out of curiosity, which iCub modules take advantage of parallel computing frameworks (e.g. CUDA, OpenCL, OpenMP, TBB)?

It would be helpful to have a brief list, in order to know which modules (from the [icub-main applications list](https://github.com/robotology/icub-main/tree/master/app)) to deploy on the Nvidia machine of our cluster and which ones on the non-Nvidia machines. For example, I know that the `motionCUT` module takes advantage of CUDA.

**Comment by pattacini** [comment, marked as answer]
`motionCUT` was used to take advantage of OpenMP while now it benefits from TBB, not CUDA at all.

[`icub-main`](https://github.com/robotology/icub-main) does not contain CUDA dependent code, if I'm not mistaken.
By contrast, repositories such as [`iol`](https://github.com/robotology/iol), [`stereo-vision`](https://github.com/robotology/stereo-vision) and [`himrep`](https://github.com/robotology/himrep) require the `SiftGPU` package and thus CUDA.

Regarding the TBB layer, I think that only `motionCUT` documentation does explicitly mention this requirement. However, it is always a good rule to enable TBB binding while compiling OpenCV.

**Comment by gsaponaro** [comment]
Thanks for the clarification.

Regarding `motionCUT`, part of the reason I got confused among the different frameworks, was that it sets an internal macro with OpenMP in its name:

```
    // check if OpenCV supports OpenMP multi-threading
    #if (CV_MAJOR_VERSION>0) && (CV_MAJOR_VERSION<3)
        #define _MOTIONCUT_MULTITHREADING_OPENMP
    #endif
```

but I guess `motionCUT` is agnostic to whatever optimization OpenCV is configured/compiled to use.

**Comment by pattacini** [comment]
Well, `motionCUT` is not completely agnostic, in a sense that it knows that some old versions of OpenCV were designed to work with OpenMP, while the most recent versions deal with the new TBB layer.

Of course, if OpenCV is not configured/compiled to use those multi-threading layers, then this condition is resolved internally to OpenCV.
